miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+                            SmtLeaf::Single((key, value))
+                        } else {
+                            // The key had an empty value before, so the leaf was empty
+                            SmtLeaf::Empty(leaf_index)
+                        }
+                    } else {
+                        // Multiple old values - filter out any with EMPTY_WORD values
+                        let entries = Vec::<(Word, Word)>::from_iter(
+                            old_key_values.into_iter().filter(|(_, v)| *v != EMPTY_WORD),
+                        );
+                        SmtLeaf::new(entries, leaf_index)
+                            .expect("consistency is given by construction")
+                    }
+                },
+                SmtLeaf::Single((current_key, current_value)) => {
+                    // Current leaf has a single entry
+                    // Check if this entry was modified
+                    let mut leaf_entries = vec![(current_key, current_value)];
+
+                    // Replace with old values
+                    for (old_key, old_value) in &old_key_values {
+                        if *old_key == current_key {
+                            // Replace the current value with the old value
+                            leaf_entries[0].1 = *old_value;
+                        } else if *old_value != EMPTY_WORD {
+                            // This key was removed in the mutation, add it back
+                            leaf_entries.push((*old_key, *old_value));
+                        }
+                    }
+
+                    // Filter out empty values and construct the appropriate leaf type
+                    let entries = Vec::<(Word, Word)>::from_iter(
+                        leaf_entries.into_iter().filter(|(_, v)| *v != EMPTY_WORD),
+                    );
+
+                    SmtLeaf::new(entries, leaf_index).expect("consistency is always given")
+                },
+                SmtLeaf::Multiple(current_entries) => {
+                    // Current leaf has multiple entries
+                    let mut leaf_entries = current_entries.clone();
+
+                    // Apply the old values
+                    for (old_key, old_value) in &old_key_values {
+                        // Find if this key exists in current entries
+                        if let Some(pos) = leaf_entries.iter().position(|(k, _)| k == old_key) {
+                            if *old_value != EMPTY_WORD {
+                                // Update the value
+                                leaf_entries[pos] = (*old_key, *old_value);
+                            } else {
+                                // Remove this entry (it didn't exist before)
+                                leaf_entries.remove(pos);
+                            }
+                        } else if *old_value != EMPTY_WORD {
+                            // This key was removed in the mutation, add it back
```

Comment:
For the constructor change, simply add: node_hashes: Arc<DashMap<NodeIndex, parking_lot::RwLock<Word>>> to InnerState. parking_lot's 40-byte locks (vs 48-byte std::sync) provide 2-5x better performance and natural boundaries for future Rayon parallelization of hash computations across tree branches.